### ISSUE TITLE
Required revision messages

### ISSUE THREAD

**Issue by ghost** [backdrop/backdrop-issues#4302]
**Description of the need**
When you have content that is quite important to keep track of changes to, you can enable revisions and turn them on by default for the content type. But there doesn't seem to be a way to 'force' editors to enter a revision message when updating content...

**Proposed solution**
I'd like to see:
- The ability to make the revision message a required field on nodes
- The ability to disable/hide the 'Create new revision' checkbox (so you can't disable it)

**Alternatives that have been considered**
Drupal has an [Enforce revision log message](https://www.drupal.org/project/enforce_revlog) module that sounds like what I'm envisioning. Perhaps that can be ported an merged into core...?

**Comment by Graham-72** [comment]
I would like us to have this.

**Comment by klonos** [comment]
I agree that this would be something useful, and something that I'd expect to be available OOTB.

**Comment by BWPanda** [comment]
Here's an initial attempt at a PR: https://github.com/backdrop/backdrop/pull/3951

It still needs some work though; e.g. choosing to create a revision and saving a node without a log message, you're shown an error and remain on the node edit form. However the 'Create new revision' checkbox is now empty (which doesn't make sense.

Happy for someone else to run with this if there's interest. Don't wait for me to come back to it 😉

**Comment by BWPanda** [comment]
We don't have a way to enforce revisions, just to enable them by default. Need a separate issue for that...

**Comment by klonos** [comment]
> However the 'Create new revision' checkbox is now empty (which doesn't make sense).

@BWPanda the culprit in this situation is this bit of code: https://github.com/backdrop/backdrop/blob/1.x/core/modules/node/node.pages.inc#L249

You see, that bit of code there is so that when you start typing in the "Revision log message" field, the "Create new revision" gets automatically ticked:

![states-magic](https://user-images.githubusercontent.com/2423362/153625439-1dc4cfc0-57d8-4f83-83f3-054e4686de88.gif)

The entire implementation of `#states` there is wrong/silly (or rather backwards or the opposite of how things should be handled). What should happen instead is this (regardless of whether the log message is set to be required or not):

![show-hide](https://user-images.githubusercontent.com/2423362/153626006-9d9ce022-5c97-41a2-8c39-1bf9ab83f3e8.gif)

Here's a PR against your branch: https://github.com/BWPanda/backdrop/pull/16 ...the changes and commits may seem many, but that's because your fork/branch have not been rebased with current backdrop 1.x. If you rebase and push, then the changes I propose will be limited to these commits:
- [BWPanda/backdrop@`e4a8c48` (#16)](https://github.com/BWPanda/backdrop/pull/16/commits/e4a8c4885b33077d33cdfc28997c332bd23e4850)
- [BWPanda/backdrop@`49e26f4` (#16)](https://github.com/BWPanda/backdrop/pull/16/commits/49e26f48730c55e48958089c906cb26ef64272d6)
- [BWPanda/backdrop@`23d89e0` (#16)](https://github.com/BWPanda/backdrop/pull/16/commits/23d89e02ec6a5a8765f31d6b549afc69ebeaeb0f)
- [BWPanda/backdrop@`87d9174` (#16)](https://github.com/BWPanda/backdrop/pull/16/commits/87d91743fc4cd11acb5a0c7859452f58a5192c61)

**Comment by klonos** [comment]
Follow-up issues:
- #5503
- upon form validation failure, automatically select the vtab that contains the required field

**Comment by BWPanda** [comment]
Thanks @klonos! PR updated.

**Comment by klonos** [comment]
Thanks @BWPanda ...the problem now is that I cannot be the one to test/review this 😅

**Comment by klonos** [comment]
FTR, while discussing #5503 and #5504 during the dev meeting today, @jenlampton suggested that we should generally be providing sensible defaults for required fields that are "hidden" inside vertical tabs (to ensure that validation errors do not happen in the first place). I could not think of a good default for the revision message field, but @jenlampton suggested a default like "Updated on [date]" or similar. That seemed like a good idea when it was mentioned, however upon further consideration I thought:

a) what's the point of providing such a message when the date and author are already provided as revision "metadata":
![image](https://user-images.githubusercontent.com/2423362/154638542-d8f5cc14-af4a-44ac-9d89-ba63bc5dc9ae.png)
b) in this specific case, that would defeat the purpose of making the revision message "required" in the first place (which is to "force" people to add a meaningful revision log message about the actual content changes that took place). That would be somewhat similar to making alt image fields required, and then providing "generic" default values. Sure, people can be "pricks" and add a dot or some rubbish text in the revision message (as some already do with required alt fields), but that's on them.

**Comment by klonos** [comment]
> ...the danger here is that people just didn't see the input field - hidden away in vertical tabs.

Yup, that is the UX issue I'm trying to address with #5503. Even if content editors chose to keep ignoring the vertical tabs that have "required" indicators (because they expect default values to have been added by the developers), we then have #5504 which will focus the vertical tab(s) that need attention.

**Comment by indigoxela** [comment]
@BWPanda I'm late here, I know, but why does this have to go into core? It's contrib in D7, with only 288 installs.

If we merge it into core, we probably also have to take care of the migration path: variable to config, mark that module as "merged into core"... Should the PR also cover that?

**Comment by BWPanda** [comment]
I disagree with the idea of giving this a default value. As stated above, it defeats the purpose. If @jenlampton is adamant that we can't put required fields into vertical tabs*, then I guess the only option is to move this field out of the vertical tab...

* However, based on the discussion in the meeting, it seems (and Jen even admitted) that she was in the minority. The majority of us thought you should be able to have required fields in vertical tabs (and even do already in certain contrib modules).

**Comment by klonos** [comment]
Disclaimer: I'm on the "majority" camp on this.

I wanted to clarify that the UX problem is not "putting required fields in vertical tabs or collapsed-by-default fieldsets" - that is absolutely fine. The actual problem manifests itself **_when such fields do not have a default value set_**.

**Comment by jenlampton** [comment]
> The majority of us thought you should be able to have required fields in vertical tabs (and even do already in certain contrib modules).

I am certain that nearly all Drupal developers were never told why vertical tabs were added to core, or even explained the various problems that were being solved by adding them.  Developers have been using them however they like ever since, and the abuse of vertical tabs is something that I have a very strong personal distaste for (as was obvious by my reaction to the other issue) mostly because of all my experience on the Drupal usability team. It's painful to watch hundreds of smart and often tech-savvy people stumble through the Drupal UI. (I may have some lingering trauma...)

Vertical tabs are an interesting and very useful creation, and it's true that they could be well suited for lots of different purposes. That said, the intent was not simply to provide an alternative to fieldsets for organizing information in a single form, it was to improve the user experience of the whole product, over time. Since nearly all developers do not consider the user experience of the whole product, over time, while building a single complicated form (I am guilty of this myself). It makes complete sense that a developer would use vertical tabs for their form when it seems like the best tool they have the problem they are trying to solve.

For Backdrop, we are generally more focused on user experience than Drupal was. This is one of the factors that we were hoping would set us apart. In core we have not misused any of the design patterns that have been provided to improve the user experience of the whole product, over time. This would be the first example of doing so. 

> @BWPanda I'm late here, I know, but why does this have to go into core? It's contrib in D7, with only 288 installs.

I agree with @indigoxela here. This is a perfect candidate for contrib considering that it does not make the 80% use rule we use for core features, in addition to presenting issues with usability.

**Comment by klonos** [comment]
>> ... It's contrib in D7, with only 288 installs.
>
> ...This is a perfect candidate for contrib considering that it does not make the 80% use rule we use for core features.

Counter-argument (not for this use case specifically, but in general): if people knew that these things exist in contrib, they may have been using them more, and that usage count could have been higher.

As for the 80% rule, we could use telemetry to prove this, but there is already a set of similar options in that same form that I bet aren't 80% either:
<img width=50% src="https://user-images.githubusercontent.com/2423362/155293743-f845991c-99a1-4710-80cc-5595235274c7.png">

I see this addition here as providing consistency of enforcement between these similar options.

**Comment by jenlampton** [comment]
> if people knew that these things exist in contrib, they may have been using them more, and that usage count could have been higher.

That argument is the same for everything in contrib, and yet most of the projects people really use have higher usage statistics. I'm uncomfortable sacrificing the editor experience for this.

**Comment by yorkshire-pudding** [comment]
I should mention that this module now exists in Backdrop contrib:
https://github.com/backdrop-contrib/enforce_revlog

Usage fairly low but I use it on several sites. There are a few improvements over the D7 version.
